### PR TITLE
Remove DiscoveryPlugin#getDiscoveryTypes

### PR DESCRIPTION
**Pull request by DaveCTurner**
With this change we no longer support pluggable discovery implementations. No
known implementations of `DiscoveryPlugin` actually override this method, so in
practice this should have no effect on the wider world. However, we were using
this rather extensively in tests to provide the `test-zen` discovery type. We
no longer need a separate discovery type for tests as we no longer need to
customise its behaviour.

Relates #38410

### CODE REVIEW

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/default-distro

File: server/src/test/java/org/elasticsearch/cluster/coordination/Zen1IT.java [hunk 1]
```diff
@@ -71,12 +71,10 @@
 public class Zen1IT extends ESIntegTestCase {
 
     private static Settings ZEN1_SETTINGS = Coordinator.addZen1Attribute(true, Settings.builder()
-        .put(TestZenDiscovery.USE_ZEN2.getKey(), false)
-        .put(TestZenDiscovery.USE_MOCK_PINGS.getKey(), false)) // Zen2 does not know about mock pings
-        .build();
+        .put(DiscoveryModule.DISCOVERY_TYPE_SETTING.getKey(), DiscoveryModule.ZEN_DISCOVERY_TYPE)).build();
 
     private static Settings ZEN2_SETTINGS = Settings.builder()
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/cluster/coordination/Zen1IT.java]: should we rename the "zen2" discovery type to "default"? This means no need for the "zen" name anymore

**Review by ywelsch — APPROVED** [review]

**Comment by DaveCTurner** [comment]
@elasticmachine retest this please now that 766f7b5bf3a896d3379999143770f548fd8c4a6a is in `6.x.

**Comment by DaveCTurner** [comment]
CI failure due to https://github.com/elastic/elasticsearch/issues/37807 - @elasticmachine please run elasticsearch-ci/2.

**Review by andrershov — APPROVED** [review]
Nice, LGTM2

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/packaging-sample

**Pull request merged**